cmd/ncproxy/hcn.go

```diff
@@ -0,0 +1,255 @@
+		if err != nil {
+			return []hcn.EndpointPolicy{}, errors.Wrap(err, "failed to marshal portname")
+		}
+		policy := hcn.EndpointPolicy{
+			Type:     hcn.PortName,
+			Settings: portPolicyJSON,
+		}
+		policies = append(policies, policy)
+	}
+
+	return policies, nil
+}
+
+func createHCNNetwork(ctx context.Context, req *ncproxygrpc.HostComputeNetworkSettings) (*hcn.HostComputeNetwork, error) {
+	// Check if the network already exists, and if so return error.
+	_, err := hcn.GetNetworkByName(req.Name)
+	if err == nil {
+		return nil, errors.Errorf("network with name %q already exists", req.Name)
+	}
+
+	policies := []hcn.NetworkPolicy{}
+	if req.SwitchName != "" {
+		// Get the layer ID from the external switch. HNS will create a transparent network for
+		// any external switch that is created not through HNS so this is what we're
+		// searching for here. If the network exists, the vSwitch with this name exists.
+		extSwitch, err := hcn.GetNetworkByName(req.SwitchName)
+		if err != nil {
+			if _, ok := err.(hcn.NetworkNotFoundError); ok {
+				return nil, errors.Errorf("no network/switch with name `%s` found", req.SwitchName)
+			}
+			return nil, errors.Wrapf(err, "failed to get network/switch with name %q", req.SwitchName)
+		}
+
+		// Get layer ID and use this as the basis for what to layer the new network over.
+		if extSwitch.Health.Extra.LayeredOn == "" {
+			return nil, errors.Errorf("no layer ID found for network %q found", extSwitch.Id)
+		}
+
+		layerPolicy := hcn.LayerConstraintNetworkPolicySetting{LayerId: extSwitch.Health.Extra.LayeredOn}
+		data, err := json.Marshal(layerPolicy)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshal layer policy")
+		}
+
+		netPolicy := hcn.NetworkPolicy{
+			Type:     hcn.LayerConstraint,
+			Settings: data,
+		}
+		policies = append(policies, netPolicy)
+	}
+
+	subnets := make([]hcn.Subnet, len(req.SubnetIpaddressPrefix))
+	for i, addrPrefix := range req.SubnetIpaddressPrefix {
+		subnet := hcn.Subnet{
+			IpAddressPrefix: addrPrefix,
+			Routes: []hcn.Route{
+				{
+					NextHop:           req.DefaultGateway,
+					DestinationPrefix: "0.0.0.0/0",
+				},
+			},
+		}
+		subnets[i] = subnet
+	}
+
+	ipam := hcn.Ipam{
+		Type:    req.IpamType.String(),
+		Subnets: subnets,
+	}
+
+	network := &hcn.HostComputeNetwork{
+		Name:     req.Name,
+		Type:     hcn.NetworkType(req.Mode.String()),
+		Ipams:    []hcn.Ipam{ipam},
+		Policies: policies,
+		SchemaVersion: hcn.SchemaVersion{
+			Major: 2,
+			Minor: 2,
+		},
+	}
+
+	network, err = network.Create()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create HNS network %q", req.Name)
+	}
+
+	return network, nil
+}
+
+func getHCNNetworkResponse(network *hcn.HostComputeNetwork) (*ncproxygrpc.HostComputeNetworkSettings, error) {
+	var (
+		ipamType                int32 = 0
+		defaultGateway                = ""
+		switchName                    = ""
+		subnetIPAddressPrefixes       = []string{}
+	)
```

Comment:
If they're all default values this can be 
```go
var (
    ipamType int32
    defaultGateway string
    switchName string
    subnetIPAddressPrefixes []string
)
```